# Open International Powerlifting Federation (IPF)

# Data Dictionary

We are given the following data dictionary.  I have added DataType to each column from my best guess at the (uncleaned) dataset.

| Column Name | Datatype | Description |
|----------|----------|----------|
| Name | String | Mandatory. The name of the lifter in UTF-8 encoding. Lifters who share the same name are distinguished by use of a `#` symbol followed by a unique number. For example, two lifters both named `John Doe` would have `Name` values `John Doe #1` and `John Doe #2` respectively. |
| Sex | String | Mandatory. The sex category in which the lifter competed, `M`, `F`, or `Mx`. Mx (pronounced *Muks*) is a gender-neutral title — like Mr and Ms — originating from the UK. It is a catch-all sex category that is particularly appropriate for non-binary lifters. The `Sex` column is defined by [crates/opltypes/src/sex.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/sex.rs). |
| Event | String | Mandatory. The type of competition that the lifter entered. The `Event` column is defined by [crates/opltypes/src/event.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/event.rs). |
| Equipment | String | Mandatory. The equipment category under which the lifts were performed. Note that this *does not mean that the lifter was actually wearing that equipment!* For example, GPC-affiliated federations do not have a category that disallows knee wraps. Therefore, all lifters, even if they only wore knee sleeves, nevertheless competed in the `Wraps` equipment category, because they were allowed to wear wraps. The `Equipment` column is defined by [crates/opltypes/src/equipment.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/equipment.rs). |
| Age | Decimal | Optional. The age of the lifter on the start date of the meet, if known. Ages can be one of two types: exact or approximate. Exact ages are given as integer numbers, for example `23`. Approximate ages are given as an integer plus `0.5`, for example `23.5`. Approximate ages mean that the lifter could be either of *two* possible ages. For an approximate age of `n + 0.5`, the possible ages are `n` or `n+1`. For example, a lifter with the given age `23.5` could be either `23` or `24` -- we don't have enough information to know. Approximate ages occur because some federations only provide us with birth year information. So another way to think about approximate ages is that `23.5` implies that the lifter turns `24` that year. The `Age` column is defined by [crates/opltypes/src/age.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/age.rs). |
| AgeClass | String | Optional. The age class in which the filter falls, for example `40-45`. These classes are based on exact age of the lifter on the day of competition. AgeClass is mostly useful because sometimes a federation will report that a lifter competed in the 50-54 divison without providing any further age information. This way, we can still tag them as 50-54, even if the `Age` column is empty. The full range available to `AgeClass` is defined by [crates/opltypes/src/ageclass.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/ageclass.rs). |
| BirthYearClass | String | Optional. The birth year class in which the filter falls, for example `40-49`. The ages in the range are the oldest possible ages for the lifter that year. For example, `40-49` means "the year the lifter turns 40 through the full year in which the lifter turns 49." `BirthYearClass` is used primarily by the IPF and by IPF affiliates. Non-IPF federations tend to use `AgeClass` instead. The full range available to `BirthYearClass` is defined by [crates/opltypes/src/birthyearclass.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/birthyearclass.rs). |
| Division | String | Optional. Free-form UTF-8 text describing the division of competition, like `Open` or `Juniors 20-23` or `Professional`. Some federations are *configured* in our database, which means that we have agreed on a limited set of division options for that federation, and we have rewritten their results to only use that set, and tests enforce that. Even still, divisions are not standardized *between* configured federations: it really is free-form text, just to provide context. Information about age should not be extracted from the `Division`, but from the `AgeClass` column. |
| BodyweightKg | Decimal | Optional. The recorded bodyweight of the lifter at the time of competition, to two decimal places. |
| WeightClassKg | String | Optional. The weight class in which the lifter competed, to two decimal places. Weight classes can be specified as a maximum or as a minimum. Maximums are specified by just the number, for example `90` means "up to (and including) 90kg." minimums are specified by a `+` to the right of the number, for example `90+` means "above (and excluding) 90kg." `WeightClassKg` is defined by [crates/opltypes/src/weightclasskg.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/weightclasskg.rs). |
| **Lift**1Kg * | Decimal | Squat1-3Kg, Bench1-3Kg, Deadlift1-3Kg. Optional. First attempts for each of squat, bench, and deadlift, respectively. Maximum of two decimal places. Negative values indicate failed attempts. Not all federations report attempt information. Some federations only report Best attempts. |
| Best3**Lift**Kg * | Decimal | Best3SquatKg, Best3BenchKg, Best3DeadliftKg. Optional. Maximum of the first three successful attempts for the lift. Rarely may be negative: that is used by some federations to report the lowest weight the lifter attempted and failed. |
| TotalKg | Decimal| Optional. Sum of `Best3SquatKg`, `Best3BenchKg`, and `Best3DeadliftKg`, if all three lifts were a success. If one of the lifts was failed, or the lifter was disqualified for some other reason, the `TotalKg` is empty. Rarely, mostly for older meets, a federation will report the total but not *any* lift information. |
| Place | String | Mandatory. The recorded place of the lifter in the given division at the end of the meet. The `Place` column is defined by [crates/opltypes/src/place.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/place.rs). |
| Dots | Decimal | Optional. A positive number if Dots points could be calculated, empty if the lifter was disqualified. Dots is very similar to (and drop-in compatible with) the original Wilks formula. It uses an updated, simpler polynomial and is built against data from drug-tested Raw lifters, as opposed to against data from drug-tested Single-ply lifters. The Dots formula was created by Tim Konertz of the BVDK in 2019. The calculation of Dots points is defined by [crates/coefficients/src/dots.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/coefficients/src/dots.rs). |
| Wilks | Decimal | Optional. A positive number if Wilks points could be calculated, empty if the lifter was disqualified. Wilks is the most common formula used for determining Best Lifter in a powerlifting meet. The calculation of Wilks points is defined by [crates/coefficients/src/wilks.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/coefficients/src/wilks.rs). |
| Glossbrenner | Decimal | Optional. A positive number if Glossbrenner points could be calculated, empty if the lifter was disqualified. Glossbrenner was created by Herb Glossbrenner as an update of the Wilks formula. It is most commonly used by GPC-affiliated federations. The calculation of Glossbrenner points is defined by [crates/coefficients/src/glossbrenner.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/coefficients/src/glossbrenner.rs). |
| Goodlift | Decimal | IPF GL Points. The successor to IPF Points (2019-01-01 through 2020-04-30). Optional. A positive number if IPF GL Points could be calculated, empty if the lifter was disqualified or IPF GL Points were undefined for the Event type. IPF GL Points roughly express relative performance to the expected performance of that weight class at an IPF World Championship event, as a percentage. The calculation of IPF GL points is defined by [crates/coefficients/src/goodlift.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/coefficients/src/goodlift.rs). |
| Tested | String | Optional. `Yes` if the lifter entered a drug-tested category, empty otherwise. Note that this records whether the results *count as drug-tested*, which does not imply that the lifter actually took a drug test. Federations do not report which lifters, if any, were subject to drug testing. |
| Country | String | Optional. The home country of the lifter, if known. The full list of valid Country values is defined by [crates/opltypes/src/country.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/country.rs). |
| State | String | Optional. The home state/province/oblast/division/etc of the lifter, if known. The full list of valid State values is defined by [crates/opltypes/src/states.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/states.rs).Expanded names are given there in comments. |
| Federation | String | Mandatory. The federation that hosted the meet. Note that this may be different than the international federation that provided sanction to the meet. For example, USPA meets are sanctioned by the IPL, but we record USPA meets as `USPA`. The full list of valid Federation values is defined by [crates/opltypes/src/federation.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/federation.rs). Comments in that file help explain what each federation value means. |
| ParentFederation | String | Optional. The topmost federation that sanctioned the meet, usually the international body. For example, the `ParentFederation` for the `USAPL` and `EPA` is `IPF`. |
| Date | Date | Mandatory. The start date of the meet in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601). ISO 8601 looks like `YYYY-MM-DD`: as an example, `1996-12-04` would be December 4th, 1996. Meets that last more than one day only have the start date recorded. |
| MeetCountry | String | Mandatory. The country in which the meet was held. The full list of valid Country values is defined by [crates/opltypes/src/country.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/country.rs). |
| MeetState | String | Optional. The state, province, or region in which the meet was held. The full list of valid State values is defined by [crates/opltypes/src/state.rs](https://gitlab.com/openpowerlifting/opl-data/blob/main/crates/opltypes/src/state.rs). |
| MeetTown | String | Undefined in the provided data dictionary. |
| MeetName | String | Mandatory. The name of the meet. The name is defined to never include the year or the federation. For example, the meet officially called `2019 USAPL Raw National Championships` would have the MeetName `Raw National Championshps`. |

\* The `LiftKg` and `Best3LiftKg` columns contain the description for several distinct columns.  In the case of `Lift1Kg`, this includes the description for Squat1Kg, Squat2Kg, Squat3Kg, Bench1Kg, Bench2Kg, Bench3Kg, Deadlift1Kg, Deadlift2Kg, and Deadlift3Kg columns.  In the case of `Best3LiftKg`, this includes the description for Best3SquatKg, Best3BenchKg, and Best3DeadliftKg columns.


## Data Cleaning

I'm working with the data from <a href='https://openpowerlifting.gitlab.io/opl-csv/bulk-csv.html'>OpenPowerlifting Data Service</a>

The purpose in this notebook is to produce a clean dataset that can be utilized for further exploration.  I want to be able to determine the answers to questions like the following:

- an assessment of different strength metrics (dots, wilks, etc)
- record progression by year
- weight class cutoffs vs records
- meet location vs lifter federation
- Patterns between weight classes and between M/F lifters
- Popularity of equipment over time (single/multi/raw/etc)
- Performance by event type (full sbd, bench only, dl only)
- Differences between Tested and Untested lifters

No doubt some of these columns will require some work before I can pull meaningful information out of them.

In [288]:
# import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

In [289]:
openpl = pd.read_csv('data/openpowerlifting-2023-08-05-ae3dc469.csv')
openpl.head()

C:\Users\Eric Bohner\AppData\Local\Temp\ipykernel_17564\579076957.py:1: DtypeWarning: Columns (33,35,38) have mixed types. Specify dtype option on import or set low_memory=False.
  openpl = pd.read_csv('data/openpowerlifting-2023-08-05-ae3dc469.csv')


Name Sex Event Equipment   Age AgeClass BirthYearClass  \
0             Alona Vladi   F   SBD       Raw  33.0    24-34          24-39   
1      Galina Solovyanova   F   SBD       Raw  43.0    40-44          40-49   
2          Daniil Voronin   M   SBD       Raw  15.5    16-17          14-18   
3          Aleksey Krasov   M   SBD       Raw  35.0    35-39          24-39   
4  Margarita Pleschenkova   M   SBD       Raw  26.5    24-34          24-39   

  Division  BodyweightKg WeightClassKg  Squat1Kg  Squat2Kg  Squat3Kg  \
0        O         58.30            60      75.0      80.0     -90.0   
1       M1         73.10            75      95.0     100.0     105.0   
2        T         67.40            75      85.0      90.0     100.0   
3        O         66.65            75     125.0     132.0     137.5   
4        O         72.45            75      80.0      85.0      90.0   

   Squat4Kg  Best3SquatKg  Bench1Kg  Bench2Kg  Bench3Kg  Bench4Kg  \
0       NaN          80.0      50.0      55.0      60.0       NaN   
1       NaN         105.0      62.5      67.5     -72.5       NaN   
2       NaN         100.0      55.0      62.5     -65.0       NaN   
3       NaN         137.5     115.0     122.5    -127.5       NaN   
4       NaN          90.0      40.0      50.0     -60.0       NaN   

   Best3BenchKg  Deadlift1Kg  Deadlift2Kg  Deadlift3Kg  Deadlift4Kg  \
0          60.0         95.0        105.0        107.5          NaN   
1          67.5        100.0        110.0       -120.0          NaN   
2          62.5         90.0        100.0        105.0          NaN   
3         122.5        150.0        165.0        170.0          NaN   
4          50.0        112.5        120.0        125.0          NaN   

   Best3DeadliftKg  TotalKg Place    Dots   Wilks  Glossbrenner  Goodlift  \
0            107.5    247.5     1  279.44  282.18        249.42     57.10   
1            110.0    282.5     1  278.95  272.99        240.35     56.76   
2            105.0    267.5     1  206.40  206.49        200.45     41.24   
3            170.0    430.0     1  334.49  334.94        325.32     66.68   
4            125.0    265.0     1  194.46  193.55        187.29     39.34   

  Tested Country State Federation ParentFederation        Date MeetCountry  \
0    Yes  Russia   NaN        GFP              NaN  2019-05-11      Russia   
1    Yes  Russia   NaN        GFP              NaN  2019-05-11      Russia   
2    Yes  Russia   NaN        GFP              NaN  2019-05-11      Russia   
3    Yes  Russia   NaN        GFP              NaN  2019-05-11      Russia   
4    Yes  Russia   NaN        GFP              NaN  2019-05-11      Russia   

  MeetState MeetTown         MeetName  
0       NaN  Bryansk  Open Tournament  
1       NaN  Bryansk  Open Tournament  
2       NaN  Bryansk  Open Tournament  
3       NaN  Bryansk  Open Tournament  
4       NaN  Bryansk  Open Tournament

In [290]:
# Let's make sure this dataset actually has the columns that are described in the data dictionary.
openpl.columns

Index(['Name', 'Sex', 'Event', 'Equipment', 'Age', 'AgeClass',
       'BirthYearClass', 'Division', 'BodyweightKg', 'WeightClassKg',
       'Squat1Kg', 'Squat2Kg', 'Squat3Kg', 'Squat4Kg', 'Best3SquatKg',
       'Bench1Kg', 'Bench2Kg', 'Bench3Kg', 'Bench4Kg', 'Best3BenchKg',
       'Deadlift1Kg', 'Deadlift2Kg', 'Deadlift3Kg', 'Deadlift4Kg',
       'Best3DeadliftKg', 'TotalKg', 'Place', 'Dots', 'Wilks', 'Glossbrenner',
       'Goodlift', 'Tested', 'Country', 'State', 'Federation',
       'ParentFederation', 'Date', 'MeetCountry', 'MeetState', 'MeetTown',
       'MeetName'],
      dtype='object')

In [291]:
# We are given a warning that certain columns have multiple datatypes.  Let's find those.

print(f'Column names with mixed data type: {openpl.columns[[33,35,38]]}')

Column names with mixed data type: Index(['State', 'ParentFederation', 'MeetState'], dtype='object')


In [292]:
openpl[['State', 'ParentFederation', 'MeetState']].sample(20)

State ParentFederation MeetState
1323969   NaN              NaN        LA
950025    NaN              NaN       NaN
718638    NaN              IPL       NaN
2179154    IL              IPL        IL
2138567    WA              IPL        WA
1382434   NaN              IPF        NM
685762    NaN              GPC       NaN
1140606   NaN              WPC        FL
2396721   NaN              IPF       VLA
1087964   NaN              IPF       NaN
1909629   NaN              IPF       KLU
2422134   NaN              UPC       NaN
876873    NaN              NaN        TX
765381    NaN              NaN        TX
837745    NaN              NaN        TX
2808285   NaN              NaN        TX
1777675   NaN              IPF       MOW
642463    NaN              IPF       NaN
1930027   NaN            WDFPF       NaN
1855316   NaN              IPF       VLA

In [293]:
# glance at duplicated rows
openpl.loc[openpl.duplicated()].head()

Name Sex Event  Equipment   Age AgeClass  \
889   Sandro Gonzaga Siqueira   M   SBD      Wraps  49.0    45-49   
9160        Mieczysłąw Maczka   M     B  Multi-ply  66.0    65-69   
9161        Mieczysłąw Maczka   M     B  Multi-ply  66.0    65-69   
9471         Fabian Piekarski   M     B        Raw  14.0    13-15   
9473      Krystian Tuzikowski   M     B        Raw  19.0    18-19   

     BirthYearClass                  Division  BodyweightKg WeightClassKg  \
889             NaN                   M 45-49         71.00            75   
9160          60-69         Pro Masters 65-69         81.80          82.5   
9161          60-69  Pro Military/Fire/Police         81.80          82.5   
9471          14-18            Pro Teen 13-15         94.25           100   
9473          19-23            Pro Teen 18-19         64.30          67.5   

      Squat1Kg  Squat2Kg  Squat3Kg  Squat4Kg  Best3SquatKg  Bench1Kg  \
889      152.5    -155.0     160.0       NaN         160.0      80.0   
9160       NaN       NaN       NaN       NaN           NaN      92.5   
9161       NaN       NaN       NaN       NaN           NaN      92.5   
9471       NaN       NaN       NaN       NaN           NaN     -90.0   
9473       NaN       NaN       NaN       NaN           NaN     115.0   

      Bench2Kg  Bench3Kg  Bench4Kg  Best3BenchKg  Deadlift1Kg  Deadlift2Kg  \
889       85.0     -90.0       NaN          85.0        170.0        180.0   
9160     100.0     102.5       NaN         102.5          NaN          NaN   
9161     100.0     102.5       NaN         102.5          NaN          NaN   
9471     -95.0    -100.0       NaN           NaN          NaN          NaN   
9473    -117.5    -117.5       NaN         115.0          NaN          NaN   

      Deadlift3Kg  Deadlift4Kg  Best3DeadliftKg  TotalKg Place    Dots  \
889         190.0       -200.0            190.0    435.0     1  323.58   
9160          NaN          NaN              NaN    102.5     1   69.77   
9161          NaN          NaN              NaN    102.5     1   69.77   
9471          NaN          NaN              NaN      NaN    DQ     NaN   
9473          NaN          NaN              NaN    115.0     1   91.89   

       Wilks  Glossbrenner  Goodlift Tested  Country State Federation  \
889   322.50        312.34     65.25    NaN      NaN   NaN    CONBRAP   
9160   69.02         66.44     36.87    NaN  Germany   NaN        WPA   
9161   69.02         66.44     36.87    NaN  Germany   NaN        WPA   
9471     NaN           NaN       NaN    NaN   Poland   NaN        WPA   
9473   92.29         89.79     66.17    NaN   Poland   NaN        WPA   

     ParentFederation        Date MeetCountry MeetState   MeetTown  \
889               GPC  2019-07-17      Brazil       NaN     Santos   
9160              NaN  2021-04-14      Poland       NaN  Pabianice   
9161              NaN  2021-04-14      Poland       NaN  Pabianice   
9471              NaN  2021-04-14      Poland       NaN  Pabianice   
9473              NaN  2021-04-14      Poland       NaN  Pabianice   

                                   MeetName  
889   Campeonato Brasileiro de Powerlifting  
9160                    World Championships  
9161                    World Championships  
9471                    World Championships  
9473                    World Championships

In [294]:
# calculate the number of duplicates as a percent of total dataset
(openpl.duplicated().sum() / len(openpl))*100

0.11897810836654735

Duplicated rows do not seem to be legitimate entries of an athlete's results.  They also only make up ~0.1% of the total dataset. For this reason, I'll simply drop them from my DataFrame.

In [295]:
# drop duplicates
openpl.drop_duplicates(inplace=True)

The `Date` column does not correctly import datetime from csv.  Let's correct that.

In [296]:
openpl['Date'] = pd.to_datetime(openpl['Date'])

Before I do any serious analysis of the data, I want to create a unique id for each athlete and for each meet/competition.  This way, athletes can be identified without having to rely on the '#x' appended to their name when two athletes have the same name.  I also want to identify meets without having to deal with the four meet address columns: `MeetCountry`, `MeetState`, `MeetTown`, and `MeetName`.

In [297]:
# adding unique athlete ids
openpl['athlete_id'], unique_names = pd.factorize(openpl['Name'])
openpl['athlete_id'] += 1 # add 1 to athlete ID

# ensure that it worked.  We want the same number of unique names as unique ids
print('Number of unique Names:', openpl['Name'].nunique())
print('Number of unique IDs for atheletes:', openpl['athlete_id'].nunique())

Number of unique Names: 772587
Number of unique IDs for atheletes: 772587


In [298]:
openpl.head()

Name Sex Event Equipment   Age AgeClass BirthYearClass  \
0             Alona Vladi   F   SBD       Raw  33.0    24-34          24-39   
1      Galina Solovyanova   F   SBD       Raw  43.0    40-44          40-49   
2          Daniil Voronin   M   SBD       Raw  15.5    16-17          14-18   
3          Aleksey Krasov   M   SBD       Raw  35.0    35-39          24-39   
4  Margarita Pleschenkova   M   SBD       Raw  26.5    24-34          24-39   

  Division  BodyweightKg WeightClassKg  Squat1Kg  Squat2Kg  Squat3Kg  \
0        O         58.30            60      75.0      80.0     -90.0   
1       M1         73.10            75      95.0     100.0     105.0   
2        T         67.40            75      85.0      90.0     100.0   
3        O         66.65            75     125.0     132.0     137.5   
4        O         72.45            75      80.0      85.0      90.0   

   Squat4Kg  Best3SquatKg  Bench1Kg  Bench2Kg  Bench3Kg  Bench4Kg  \
0       NaN          80.0      50.0      55.0      60.0       NaN   
1       NaN         105.0      62.5      67.5     -72.5       NaN   
2       NaN         100.0      55.0      62.5     -65.0       NaN   
3       NaN         137.5     115.0     122.5    -127.5       NaN   
4       NaN          90.0      40.0      50.0     -60.0       NaN   

   Best3BenchKg  Deadlift1Kg  Deadlift2Kg  Deadlift3Kg  Deadlift4Kg  \
0          60.0         95.0        105.0        107.5          NaN   
1          67.5        100.0        110.0       -120.0          NaN   
2          62.5         90.0        100.0        105.0          NaN   
3         122.5        150.0        165.0        170.0          NaN   
4          50.0        112.5        120.0        125.0          NaN   

   Best3DeadliftKg  TotalKg Place    Dots   Wilks  Glossbrenner  Goodlift  \
0            107.5    247.5     1  279.44  282.18        249.42     57.10   
1            110.0    282.5     1  278.95  272.99        240.35     56.76   
2            105.0    267.5     1  206.40  206.49        200.45     41.24   
3            170.0    430.0     1  334.49  334.94        325.32     66.68   
4            125.0    265.0     1  194.46  193.55        187.29     39.34   

  Tested Country State Federation ParentFederation       Date MeetCountry  \
0    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
1    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
2    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
3    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
4    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   

  MeetState MeetTown         MeetName  athlete_id  
0       NaN  Bryansk  Open Tournament           1  
1       NaN  Bryansk  Open Tournament           2  
2       NaN  Bryansk  Open Tournament           3  
3       NaN  Bryansk  Open Tournament           4  
4       NaN  Bryansk  Open Tournament           5

In [299]:
# Create a unique meet_id based on address and meet name

# get relevant columns for meet details
meet_details = ['Date', 'MeetCountry', 'MeetState', 'MeetTown', 'MeetName']

# set a unique id for each row where the meet_details are the same
openpl.insert(loc=0, column='meet_id', value=openpl.set_index(meet_details).index.factorize()[0]+1)

In [300]:
# reorganize the columns in the order I want them
openpl = openpl[['meet_id', 'athlete_id', 'Name', 'Sex', 'Event', 'Equipment', 'Age', 'AgeClass',
       'BirthYearClass', 'Division', 'BodyweightKg', 'WeightClassKg',
       'Squat1Kg', 'Squat2Kg', 'Squat3Kg', 'Squat4Kg', 'Best3SquatKg',
       'Bench1Kg', 'Bench2Kg', 'Bench3Kg', 'Bench4Kg', 'Best3BenchKg',
       'Deadlift1Kg', 'Deadlift2Kg', 'Deadlift3Kg', 'Deadlift4Kg',
       'Best3DeadliftKg', 'TotalKg', 'Place', 'Dots', 'Wilks', 'Glossbrenner',
       'Goodlift', 'Tested', 'Country', 'State', 'Federation',
       'ParentFederation', 'Date', 'MeetCountry', 'MeetState', 'MeetTown',
       'MeetName']]

In [301]:
openpl.head()

meet_id  athlete_id                    Name Sex Event Equipment   Age  \
0        1           1             Alona Vladi   F   SBD       Raw  33.0   
1        1           2      Galina Solovyanova   F   SBD       Raw  43.0   
2        1           3          Daniil Voronin   M   SBD       Raw  15.5   
3        1           4          Aleksey Krasov   M   SBD       Raw  35.0   
4        1           5  Margarita Pleschenkova   M   SBD       Raw  26.5   

  AgeClass BirthYearClass Division  BodyweightKg WeightClassKg  Squat1Kg  \
0    24-34          24-39        O         58.30            60      75.0   
1    40-44          40-49       M1         73.10            75      95.0   
2    16-17          14-18        T         67.40            75      85.0   
3    35-39          24-39        O         66.65            75     125.0   
4    24-34          24-39        O         72.45            75      80.0   

   Squat2Kg  Squat3Kg  Squat4Kg  Best3SquatKg  Bench1Kg  Bench2Kg  Bench3Kg  \
0      80.0     -90.0       NaN          80.0      50.0      55.0      60.0   
1     100.0     105.0       NaN         105.0      62.5      67.5     -72.5   
2      90.0     100.0       NaN         100.0      55.0      62.5     -65.0   
3     132.0     137.5       NaN         137.5     115.0     122.5    -127.5   
4      85.0      90.0       NaN          90.0      40.0      50.0     -60.0   

   Bench4Kg  Best3BenchKg  Deadlift1Kg  Deadlift2Kg  Deadlift3Kg  Deadlift4Kg  \
0       NaN          60.0         95.0        105.0        107.5          NaN   
1       NaN          67.5        100.0        110.0       -120.0          NaN   
2       NaN          62.5         90.0        100.0        105.0          NaN   
3       NaN         122.5        150.0        165.0        170.0          NaN   
4       NaN          50.0        112.5        120.0        125.0          NaN   

   Best3DeadliftKg  TotalKg Place    Dots   Wilks  Glossbrenner  Goodlift  \
0            107.5    247.5     1  279.44  282.18        249.42     57.10   
1            110.0    282.5     1  278.95  272.99        240.35     56.76   
2            105.0    267.5     1  206.40  206.49        200.45     41.24   
3            170.0    430.0     1  334.49  334.94        325.32     66.68   
4            125.0    265.0     1  194.46  193.55        187.29     39.34   

  Tested Country State Federation ParentFederation       Date MeetCountry  \
0    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
1    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
2    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
3    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   
4    Yes  Russia   NaN        GFP              NaN 2019-05-11      Russia   

  MeetState MeetTown         MeetName  
0       NaN  Bryansk  Open Tournament  
1       NaN  Bryansk  Open Tournament  
2       NaN  Bryansk  Open Tournament  
3       NaN  Bryansk  Open Tournament  
4       NaN  Bryansk  Open Tournament

In [302]:
# Find some basic information about athletes and meets:
print('The number of unique athletes:', openpl['athlete_id'].nunique())
print('The number of unique meets:', openpl['meet_id'].nunique())

The number of unique athletes: 772587
The number of unique meets: 44991


Now that I have a meet_id and athlete_id, I can start to find out information about them.  Above, we see that our dataset with ~3 million rows has ~772,000 athletes competing in ~45,000 different competitions.  That's a good start, but now we want to find out other useful information.  Many questions I want to find the answers to involve comparing different weight classes.  So, let's look at the weight class column to see what we can find.

In [303]:
# find the number of different weight classes
openpl['WeightClassKg'].nunique()

388

In [304]:
openpl['WeightClassKg'].unique()

array(['60', '75', '85', '105', '55', '95', '135+', '51', '70', '90+',
       '120', '120+', '45', '65', '135', '82.5', '140', '100', '110',
       '125', '90', '140+', '67.5', '52', '56', '110+', '125+', '48',
       '82.5+', '44', nan, '57', '63', '72', '84', '84+', '59', '66',
       '74', '83', '93', '47', '53', '64', '94', '69', '77', '58', '62',
       '105+', '100+', '155', '155+', '43', '35', '44.5', '30', '40',
       '145', '145+', '76', '63+', '118', '80', '130', '130+', '39',
       '50.5', '55.5', '58.5', '47.5', '61', '82', '109', '109+', '95.2',
       '33', '+', '75+', '67.5+', '67', '70+', '50', '115', '115+',
       '51.7', '55.7', '59.8', '65.7', '70.3', '74.8', '82.1', '88',
       '93.8', '99.7', '109.7', '124.7', '124.7+', '47.6', '89.8',
       '99.7+', '67.1', '80+', '60+', '57+', '83+', '74+', '72+', '69+',
       '93+', '76+', '79.8', '102.5', '125.1', '125.1+', '79.8+', '113.8',
       '113.8+', '73', '91.1+', '91.1', '63.9', '136.5', '136.5+', '68.4',
      

There are 388 unique weight classes, but I am seeing a lot of weight classes that don't make sense.  For instance, '+' and '1+' are not weight classes.  But there's a chance competitors assigned to those classes make a significant difference to the stats we care about, like records.  So, I want to avoid simply dropping those rows.  Instead, I'll assign weight classes according to the weight classes that have been used by the IPF, USAPL, SPF, and USPA:

#### Classes and categories (Copied from Wikipedia):
Weight Classes:

Most powerlifting federations use the following weight classes:

- Men: 52 kg, 56 kg, 60 kg, 67.5 kg, 75 kg, 82.5 kg, 90 kg, 100 kg, 110 kg, 125 kg, 140 kg, 140 kg+
- Women: 44 kg, 48 kg, 52 kg, 56 kg, 60 kg, 67.5 kg, 75 kg, 82.5 kg, 90 kg, 90 kg+

However, in 2011, the IPF introduced the following new weight classes:

IPF Weight Classes:

- Men: up to 53 kg (Sub-Junior/Junior), 59 kg, 66 kg, 74 kg, 83 kg, 93 kg, 105 kg, 120 kg, 120 kg+
- Women: up to 43 kg (Sub-Junior/Junior), 47 kg, 52 kg, 57 kg, 63 kg, 69 kg, 76 kg, 84 kg, 84 kg+

In [305]:
# defining weight classes

# ipf after 2011 rule changes.  The 120+ and 84+ categories are converted to 1200 and 840 for classification purposes
ipf_2011m = [53, 59, 66, 74, 83, 93, 105, 120, 1200]
ipf_2011w = [43, 47, 52, 57, 63, 69, 76, 84, 840]
ipf_2011mx = [] # It isn't clear what weight classes to use for gender neutral athletes

# all other weight classes, including pre-2011 ipf. Again, + categories append a 0 for classification purposes
non_ipfm = [52, 56, 60, 67.5, 75, 82.5, 90, 100, 110, 125, 140, 1400]
non_ipfw = [44, 48, 52, 56, 60, 67.5, 75, 82.5, 90, 900]
non_ipfmx = [] # It isn't clear what weight classes to use for gender neutral athletes

The "+" categories create a significant challenge.  A woman competing in the IPF who weighs 85kg after 2011, for instance, should be classified as an 84+ competitor and not a 90 competitor.  Note, also, that these categories are strings and not floats, which poses additional challenges in correctly categorizing athletes.  

Finally, "Mx" competitors (non-binary or otherwise excluded from competing in women's competitions according to certain federations' rules) will also be difficult to accurately categorize into their nearest category since some federations allow Trans Women to compete in women's categories (notably the IPF) while others do not (notably USAPL).  

For simplicity, I will assign 'Mx' competitors to male weight classes, although this creates the potential for inaccuracies.  I will also try to assign weight classes, wherever possible, according to body weight and ipf/date combinations.

In [306]:
# Find number of rows where both body weight and weight class are missing. These rows can't add meaningful information.
(openpl['BodyweightKg'].isna() & openpl['WeightClassKg'].isna()).sum()

18110

In [307]:
# dropping those 18,000 columns

# check number or rows before deletion
num_before = len(openpl)

openpl.dropna(subset=['BodyweightKg', 'WeightClassKg'], how='all', inplace=True)

print(f'Number of rows deleted: {num_before - len(openpl)}')

Number of rows deleted: 18110


In [308]:
# check to see that IPF is always correct assigned to the ParentFederation column
fed_par_ipf = ((openpl['Federation'] == 'IPF') & (openpl['ParentFederation'] == 'IPF')).sum()
fed_ipf = (openpl['Federation'] == 'IPF').sum()

print(f'IPF Federation count: {fed_ipf}. IPF ParentFederation and Federation: {fed_par_ipf}')

IPF Federation count: 46704. IPF ParentFederation and Federation: 46704


In [309]:
openpl['BodyweightKg'].dtypes

dtype('float64')

In [310]:
# find weight classes where there is no bodyweight
openpl.loc[openpl['BodyweightKg'].isna(), 'WeightClassKg'].unique()

array(['140+', '125+', '82.5+', '90+', '145+', '110+', '100+', '84+',
       '120+', '93+', '91.1+', '125.1+', '90.7+', '105+', '117.7+',
       '90.2+', '107.9+', '136+', '85+', '107+', '86+', '57+', '83+',
       '81.6+', '104.7+', '+', '102+', '81.5+', '79.3+', '60+', '80+',
       '104.3+', '144.5+', '99.7+', '72.5+', '67.5+', '75+', '91+',
       '92.9+', '68+', '56+', '117.9+', '95.2+', '63.5+', '64.4+',
       '113.4+', '58.5+', '127+', '83.9+', '70+', '61.2+', '90.5+',
       '54.4+', '88.9+', '143+', '72.6+', '96+', '78.9+', '56.7+',
       '58.9+', '101.5+', '88.4+', '275+', '155+', '124.7+', '139.7+'],
      dtype=object)

All of the weight classes where there is no bodyweight are strings with a '+' at the end.  I want to use the athlete's body weight (along with the federation and date to determine the appropriate weight class list) to find the correct weight class.  When the body weight is not available, I then want to use the weight class and correct it to one of the standard weight classes.  To do this, I can strip the '+' from the weight class and find the next highest weight class from the appropriate list.  Note that I've assigned the heavy end of the weight classes to 840, 900, etc in place of 84+, 90+, etc., so that everything remains a float and atheletes aren't accidentally sorted into, for instance, the -84 class.

In [311]:
# round up to nearest value in the weight class list
def round_up(bw_value, wc_value, wc_list):
    if pd.isna(bw_value):
        # remove '+' from wc and cast as a float
        try:
            wc_value = float(wc_value.rstrip('+'))
            closest_value = min(x for x in wc_list if x > wc_value)
        except:
            closest_value = 0
    else:
        # get the next highest value in the weight class list
        closest_value = min(x for x in wc_list if x > bw_value)
    return closest_value

# replace the weight class (wc) according to athlete's body weight (bw)
def get_wc_from_bw(row):
    if row['ParentFederation'] == 'IPF' and row['Date'].year >= 2011:
        weight_list = ipf_2011w if row['Sex'] == 'F' else ipf_2011m
    else:
        weight_list = non_ipfw if row['Sex'] == 'F' else non_ipfm
    return round_up(row['BodyweightKg'], row['WeightClassKg'], weight_list)

In [312]:
# replace weight class values from the list
openpl['WeightClassKg'] = openpl.apply(get_wc_from_bw, axis=1)

In [313]:
# See which weight classes are left


num_wc = openpl['WeightClassKg'].nunique()
weight_classes = sorted(openpl['WeightClassKg'].unique())

print(f'Number of weight classes after reclassifying: {num_wc}')
print('Weight classes:')
print(weight_classes)

print('Weight Class categories:')
print(sorted(ipf_2011m + ipf_2011w + non_ipfw + non_ipfm))


Number of weight classes after reclassifying: 33
Weight classes:
[0.0, 43.0, 44.0, 47.0, 48.0, 52.0, 53.0, 56.0, 57.0, 59.0, 60.0, 63.0, 66.0, 67.5, 69.0, 74.0, 75.0, 76.0, 82.5, 83.0, 84.0, 90.0, 93.0, 100.0, 105.0, 110.0, 120.0, 125.0, 140.0, 840.0, 900.0, 1200.0, 1400.0]
Weight Class categories:
[43, 44, 47, 48, 52, 52, 52, 53, 56, 56, 57, 59, 60, 60, 63, 66, 67.5, 67.5, 69, 74, 75, 75, 76, 82.5, 82.5, 83, 84, 90, 90, 93, 100, 105, 110, 120, 125, 140, 840, 900, 1200, 1400]


Note that, while there are 40 weight classes total, there is overlap between M and F weight classes for the non-ipf weightclasses, so we can expect to see that those duplicates disappear when looking at unique weight classes after classification.  In addition, there is now a '0' weight class where the athlete had no bodyweight and only '+' entered into the weight class category.  This number is meaningless to us, so will be dropped.

In [314]:
# look at what we are dropping
openpl.loc[openpl['WeightClassKg'] == 0]

meet_id  athlete_id             Name Sex Event Equipment   Age  \
1026565    18443      306516     Hari Fafutis   M     S       Raw  28.0   
1026601    18445      306543  Sadie Fishburne   F   SBD       Raw  17.0   
1026635    18445      306436       Luke Mabey   M   SBD       Raw  18.0   
1026887    18450      306644  Emma Greathouse   F   SBD       Raw  15.0   
1026921    18450      306512  Dylan Fleetwood   M   SBD       Raw  15.0   
1026922    18450      306656   Cooper Carroll   M   SBD       Raw  15.0   

        AgeClass BirthYearClass Division  BodyweightKg  WeightClassKg  \
1026565    24-34          24-39     Open           NaN            0.0   
1026601    16-17          14-18  Juniors           NaN            0.0   
1026635    18-19          19-23  Juniors           NaN            0.0   
1026887    13-15          14-18  Juniors           NaN            0.0   
1026921    13-15          14-18  Juniors           NaN            0.0   
1026922    13-15          14-18  Juniors           NaN            0.0   

         Squat1Kg  Squat2Kg  Squat3Kg  Squat4Kg  Best3SquatKg  Bench1Kg  \
1026565    240.00    250.00       NaN       NaN        250.00       NaN   
1026601     74.84     79.38     81.65       NaN         81.65     38.56   
1026635    163.29    170.10       NaN       NaN        170.10    113.40   
1026887    111.13    121.11    125.65       NaN        125.65    -52.16   
1026921    131.54       NaN    137.44       NaN        137.44     70.31   
1026922       NaN     43.09       NaN       NaN         43.09     43.09   

         Bench2Kg  Bench3Kg  Bench4Kg  Best3BenchKg  Deadlift1Kg  Deadlift2Kg  \
1026565       NaN       NaN       NaN           NaN          NaN          NaN   
1026601     40.82     41.96       NaN         41.96        74.84        79.38   
1026635       NaN       NaN       NaN        113.40       195.04          NaN   
1026887    -61.23    -66.68       NaN           NaN       129.27      -142.88   
1026921     77.11     84.37       NaN         84.37       131.54       149.69   
1026922       NaN       NaN       NaN         43.09        83.91          NaN   

         Deadlift3Kg  Deadlift4Kg  Best3DeadliftKg  TotalKg Place  Dots  \
1026565          NaN          NaN              NaN   250.00     1   NaN   
1026601        83.91          NaN            83.91   207.52     1   NaN   
1026635          NaN          NaN           195.04   478.54     1   NaN   
1026887          NaN          NaN           129.27      NaN    DQ   NaN   
1026921      -158.76          NaN           149.69   371.49     1   NaN   
1026922          NaN          NaN            83.91   170.10     2   NaN   

         Wilks  Glossbrenner  Goodlift Tested Country State Federation  \
1026565    NaN           NaN       NaN    NaN     NaN   NaN       USSF   
1026601    NaN           NaN       NaN    NaN     USA    MD       USSF   
1026635    NaN           NaN       NaN    NaN     USA    NJ       USSF   
1026887    NaN           NaN       NaN    NaN     USA    KY       USSF   
1026921    NaN           NaN       NaN    NaN     USA    MO       USSF   
1026922    NaN           NaN       NaN    NaN     USA    MO       USSF   

        ParentFederation       Date MeetCountry MeetState MeetTown  \
1026565              NaN 2021-10-09         USA       NaN      NaN   
1026601              NaN 2022-02-20         USA       NaN      NaN   
1026635              NaN 2022-02-20         USA       NaN      NaN   
1026887              NaN 2021-06-27         USA       NaN      NaN   
1026921              NaN 2021-06-27         USA       NaN      NaN   
1026922              NaN 2021-06-27         USA       NaN      NaN   

                              MeetName  
1026565     Powerlifting Championships  
1026601  Powerlifting Winter Challenge  
1026635  Powerlifting Winter Challenge  
1026887               Summer Challenge  
1026921               Summer Challenge  
1026922               Summer Challenge

Without knowing their weight class, Dots, Wilks, GB, or GL scores, we can't say anything about their strength metrics, so these values are useless to us.  I drop them with a free conscience. 

In [315]:
# removing the 0 weight class
openpl = openpl.loc[openpl['WeightClassKg'] != 0]

Now that we have a weight class for everyone according to their correct federation, a unique athlete ID, and a unique meet ID, we can finally focus on the actual lifting.  Here there are two goals: fill in the missing `Best3LiftKg` columns from the best of their lifts and then calculate a total from those lifts.  Following this, we can finally start to compare strength across weight categories, record progressions, lifting categories, etc.

Occassionally, lifters will bomb out, resulting in 3 (or 4) negative numbers for their attempts.  In those cases, the athlete's total may be less than their best lift (e.g., a lifter has 100, 104, 106, NaN for squat attempts, but -55, -55, -55, NaN for bench attempts, their total will be 51).  For simplicity, I'll replace NaN values with 0 for the time being.  Later, I replace those NaN values.

In [316]:
# adding np.nan to an int or float results in np.nan.  So, we'll fill the NaN values in the attempt columns temporarily
# Specify the columns to fill with 0
columns_to_fill = ['Squat1Kg', 'Squat2Kg', 'Squat3Kg', 'Squat4Kg', 'Best3SquatKg', 
                   'Bench1Kg', 'Bench2Kg', 'Bench3Kg', 'Bench4Kg', 'Best3BenchKg', 
                   'Deadlift1Kg', 'Deadlift2Kg', 'Deadlift3Kg', 'Deadlift4Kg', 'Best3DeadliftKg', 'TotalKg']

# Fill empty cells with 0 in the specified columns
openpl[columns_to_fill] = openpl[columns_to_fill].fillna(0)

In [322]:
# Find best lifts and total
def fill_best_lifts(row):

    # get best lifts from max of 4 attempts
    best_squat = max(row['Squat1Kg'], row['Squat2Kg'], row['Squat3Kg'], row['Squat4Kg'])
    best_bench = max(row['Bench1Kg'], row['Bench2Kg'], row['Bench3Kg'], row['Bench4Kg'])
    best_deadlift = max(row['Deadlift1Kg'], row['Deadlift2Kg'], row['Deadlift3Kg'], row['Deadlift4Kg'])

    # if attempts were not recorded, keep recorded best attempts and/or totals
    if row['Best3SquatKg'] > best_squat:
        best_squat = row['Best3SquatKg']
    
    if row['Best3BenchKg'] > best_bench:
        best_bench = row['Best3BenchKg']

    if row['Best3DeadliftKg'] > best_deadlift:
        best_deadlift = row['Best3DeadliftKg']

    # Calculate total
    total_kg = best_squat + best_bench + best_deadlift
    # it Total was entered and nothing else was entered, use that instead.
    if row['TotalKg'] > total_kg:
        total_kg = row['TotalKg']

    return best_squat, best_bench, best_deadlift, total_kg

In [323]:
# impute missing data
openpl[['Best3SquatKg', 'Best3BenchKg', 'Best3DeadliftKg', 'TotalKg']] = openpl.apply(fill_best_lifts, axis=1, result_type='expand')

In [324]:
# sneak a peek at what we've done
openpl.sample(10)

meet_id  athlete_id                   Name Sex Event   Equipment  \
2642518    42678      682883          Chris Marquez   M   SBD  Single-ply   
1256168    22551      387042               Paul Lee   M   SBD  Single-ply   
2337170    38845      318447      Francesco Liberti   M   SBD         Raw   
2774393    43709      695905          Caden Forisha   M   SBD  Single-ply   
1585492    27823      483565              I. Musaev   M     B         Raw   
1324134    23681      365672             D. Pittman   M   SBD  Single-ply   
1734916    30323      524714  Evgeniya Lemenovskaya   F     B  Single-ply   
2106989    35224      336139           Bob Jacobson   M   SBD         Raw   
2032731    34077      592643            Alan Torres   M    BD         Raw   
795903     15304      243275       Lilliane Jackson   F   SBD  Single-ply   

          Age AgeClass BirthYearClass       Division  BodyweightKg  \
2642518  14.5    13-15          14-18           Boys        102.06   
1256168   NaN      NaN            NaN           Open         75.00   
2337170  22.5    20-23          19-23        Juniors         68.35   
2774393   NaN      NaN            NaN           Boys         66.68   
1585492  17.5    18-19          14-18           Teen         59.00   
1324134   NaN    45-49            NaN  Masters 40-49         75.00   
1734916  21.5    20-23          19-23           Open         58.50   
2106989  61.0    60-64          60-69  Masters 60-64         96.70   
2032731  19.0    18-19          19-23  Juniors 18-19         89.40   
795903    NaN      NaN            NaN          Girls         57.97   

         WeightClassKg  Squat1Kg  Squat2Kg  Squat3Kg  Squat4Kg  Best3SquatKg  \
2642518          110.0       0.0       0.0       0.0       0.0        204.12   
1256168           82.5       0.0       0.0       0.0       0.0        174.63   
2337170           75.0     170.0     182.0     190.0       0.0        190.00   
2774393           67.5       0.0       0.0       0.0       0.0        142.88   
1585492           66.0       0.0       0.0       0.0       0.0          0.00   
1324134           82.5       0.0       0.0       0.0       0.0        192.50   
1734916           60.0       0.0       0.0       0.0       0.0          0.00   
2106989          100.0       0.0       0.0       0.0       0.0        115.00   
2032731           90.0       0.0       0.0       0.0       0.0          0.00   
795903            60.0       0.0       0.0       0.0       0.0         86.18   

         Bench1Kg  Bench2Kg  Bench3Kg  Bench4Kg  Best3BenchKg  Deadlift1Kg  \
2642518       0.0       0.0       0.0       0.0        111.13          0.0   
1256168       0.0       0.0       0.0       0.0        136.08          0.0   
2337170     109.0     115.0     118.0       0.0        118.00        182.0   
2774393       0.0       0.0       0.0       0.0         97.52          0.0   
1585492      70.0     -77.5      77.5       0.0         77.50          0.0   
1324134       0.0       0.0       0.0       0.0        145.00          0.0   
1734916      70.0     -75.0      75.0       0.0         75.00          0.0   
2106989       0.0       0.0       0.0     137.5        137.50          0.0   
2032731     115.0    -125.0    -125.0       0.0        115.00        227.5   
795903        0.0       0.0       0.0       0.0         58.97          0.0   

         Deadlift2Kg  Deadlift3Kg  Deadlift4Kg  Best3DeadliftKg  TotalKg  \
2642518          0.0          0.0          0.0           195.04   510.29   
1256168          0.0          0.0          0.0           188.24   498.95   
2337170        195.0        205.0          0.0           205.00   513.00   
2774393          0.0          0.0          0.0           170.10   410.50   
1585492          0.0          0.0          0.0             0.00    77.50   
1324134          0.0          0.0          0.0           227.50   565.00   
1734916          0.0          0.0          0.0             0.00    75.00   
2106989          0.0          0.0         

In [327]:
# check to see that totals less than a max lift always result in DQ or equivalent
condition = (
    (openpl['TotalKg'] < openpl[['Best3DeadliftKg', 'Best3BenchKg', 'Best3SquatKg']].max(axis=1))
)

# Filter the DataFrame based on the condition
df = openpl[condition]
df

meet_id  athlete_id           Name Sex Event Equipment   Age AgeClass  \
904534    16870      272642  Igor Vorobyev   M   SBD     Wraps  24.0    24-34   

       BirthYearClass Division  BodyweightKg  WeightClassKg  Squat1Kg  \
904534          24-39     Open         100.0          110.0     320.0   

        Squat2Kg  Squat3Kg  Squat4Kg  Best3SquatKg  Bench1Kg  Bench2Kg  \
904534    -335.0     345.0    -355.0         345.0     175.0     185.0   

        Bench3Kg  Bench4Kg  Best3BenchKg  Deadlift1Kg  Deadlift2Kg  \
904534    -195.0    -195.0         185.0       -330.0       -345.0   

        Deadlift3Kg  Deadlift4Kg  Best3DeadliftKg  TotalKg Place  Dots  Wilks  \
904534       -345.0       -345.0           -330.0    200.0    DQ   NaN    NaN   

        Glossbrenner  Goodlift Tested Country State Federation  \
904534           NaN       NaN    NaN  Russia   NaN       WRPF   

       ParentFederation       Date MeetCountry MeetState MeetTown  \
904534             WRPF 2018-10-04      Russia       MOW   Moscow   

                   MeetName  
904534  World Championships

A single lifter's total was incorrectly classified for one of his meets.  This is a result of that lifter taking all of their attempts (in this case 12), but failing to put up a total as a result of bombing out on deadlifts.  So, the total is "correctly" calculated as 345 + 185 + (-330) = 200.  Since this athlete was disqualified from the competition, their individual lifts (even the ones that were successful) will not count toward records or a total.  This happens to be a special case in that the athlete had 4 attempts for each lift, whereas 3 attempts is the norm.  Had there only been 3 attempts, the 4th attempt would have been a '0', which would have resulted in a total of 530.  Since this single instance does not affect anything, it can be ignored.

# Exploratory Data Analysis

Here are some interesting things I want to know:

- How many athletes are in this database?
- Who holds the S/B/D/Total records for each weight class and each federation?
- Whose record was held for the longest time?

and also these questions from above:

- an assessment of different strength metrics (dots, wilks, etc)
- record progression by year
- weight class cutoffs vs records
- meet location vs lifter federation
- Patterns between weight classes and between M/F lifters
- Popularity of equipment over time (single/multi/raw/etc)
- Performance by event type (full sbd, bench only, dl only)
- Differences between Tested and Untested lifters

In [332]:
# Find the number of athletes:

openpl['athlete_id'].nunique()

768409

In [337]:
# How many people have competed in each weight class?
athletes_by_wc = openpl.groupby('WeightClassKg')['athlete_id'].nunique().sort_values(ascending=False)
athletes_by_wc

WeightClassKg
82.5      118599
90.0      116288
75.0      112197
100.0     103230
67.5       90059
110.0      80433
125.0      59108
60.0       52036
83.0       42690
93.0       42526
56.0       38943
52.0       33593
140.0      32364
105.0      32292
74.0       30958
120.0      18269
66.0       17419
63.0       16567
900.0      16273
1400.0     12615
76.0       12521
57.0       12496
69.0       12132
1200.0      9805
48.0        9072
59.0        8775
84.0        8328
840.0       8234
53.0        4444
44.0        3992
47.0        3591
43.0        1268
Name: athlete_id, dtype: int64

In [341]:
openpl['Equipment'].unique()

array(['Raw', 'Wraps', 'Multi-ply', 'Single-ply', 'Unlimited', 'Straps'],
      dtype=object)

In [342]:
openpl['Equipment'].value_counts()

Single-ply    1319300
Raw           1288288
Wraps          194368
Multi-ply      122206
Unlimited        8482
Straps             50
Name: Equipment, dtype: int64

In [338]:
# For importing into a MySQL database, the empty cells appear as an empty string ('') rather than a NULL value.
# So, I'll replace NaNs with \N to resolve this issue.

# first replace the 0 values in the attempts back to NaNs.  This is so that averages are not skewed by a 0.
columns_to_replace = ['Squat1Kg', 'Squat2Kg', 'Squat3Kg', 'Squat4Kg', 'Best3SquatKg', 
                      'Bench1Kg', 'Bench2Kg', 'Bench3Kg', 'Bench4Kg', 'Best3BenchKg', 
                      'Deadlift1Kg', 'Deadlift2Kg', 'Deadlift3Kg', 'Deadlift4Kg', 'Best3DeadliftKg', 'TotalKg']
openpl[columns_to_replace] = openpl[columns_to_replace].replace(0, np.nan)

# replace NaNs with \N
openpl.replace(np.nan, '\\N', inplace=True)

In [340]:
# I'll write this to a csv for use later.
openpl.to_csv('openpl.csv', index=False)